subsys/lorawan/services/frag_decoder_lowmem.c

```diff
@@ -0,0 +1,411 @@
+
+static bool triangular_matrix_get_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m)
+{
+	/* We are dealing with triangular matrices, so we don't expect actions in the lower half */
+	__ASSERT(x >= y, "x: %d, y: %d, m: %d", x, y, m);
+	size_t bit;
+	int ret;
+
+	ret = sys_bitarray_test_bit(m2tbm, matrix_location_to_index(x, y, m), &bit);
+	__ASSERT_NO_MSG(ret == 0);
+
+	return bit != 0;
+}
+
+static void triangular_matrix_set_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m)
+{
+	/* We are dealing with triangular matrices, so we don't expect actions in the lower half */
+	__ASSERT(x >= y, "x: %d, y: %d, m: %d", x, y, m);
+	int ret;
+
+	ret = sys_bitarray_set_bit(m2tbm, matrix_location_to_index(x, y, m));
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static void triangular_matrix_clear_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m)
+{
+	/* We are dealing with triangular matrices, so we don't expect actions in the lower half */
+	__ASSERT(x >= y, "x: %d, y: %d, m: %d", x, y, m);
+
+	int ret;
+
+	ret = sys_bitarray_clear_bit(m2tbm, matrix_location_to_index(x, y, m));
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static inline bool bit_get(struct sys_bitarray *bitmap, size_t index)
+{
+	int bit, ret;
+
+	ret = sys_bitarray_test_bit(bitmap, index, &bit);
+	__ASSERT_NO_MSG(ret == 0);
+	return bit != 0;
+}
+
+static inline void bit_set(struct sys_bitarray *bitmap, size_t index)
+{
+	int ret;
+
+	ret = sys_bitarray_set_bit(bitmap, index);
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static inline void bit_clear(struct sys_bitarray *bitmap, size_t index)
+{
+	int ret;
+
+	ret = sys_bitarray_clear_bit(bitmap, index);
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static inline size_t bit_count_ones(struct sys_bitarray *bitmap, size_t index)
+{
+	size_t count;
+	int ret;
+
+	ret = sys_bitarray_popcount_region(bitmap, index + 1, 0, &count);
+	__ASSERT_NO_MSG(ret == 0);
+	return count;
+}
+
+static inline void bit_xor(struct sys_bitarray *des, struct sys_bitarray *src, size_t size)
+{
+	int ret;
+
+	ret = sys_bitarray_xor(des, src, size, 0);
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static inline void bit_clear_all(struct sys_bitarray *bitmap, size_t size)
+{
+	int ret;
+
+	ret = sys_bitarray_clear_region(bitmap, size, 0);
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+static int32_t prbs23(int32_t seed)
+{
+	int32_t b0 = seed & 1;
+	int32_t b1 = (seed & 32) / 32;
+
+	return (seed / 2) + ((b0 ^ b1) << 22);
+}
+
+static void frag_dec_parity_matrix_vector(size_t m, size_t n, struct sys_bitarray *vec)
+{
+	size_t mm, r;
+	int32_t x;
+	int ret;
+
+	ret = sys_bitarray_clear_region(vec, m, 0);
+	__ASSERT_NO_MSG(ret == 0);
+
+	/*
+	 * Powers of 2 must be treated differently to make sure matrix content is close
+	 * to random. Powers of 2 tend to generate patterns.
+	 */
+	if (is_power_of_two(m)) {
+		mm = m + 1;
+	} else {
+		mm = m;
+	}
+
+	x = 1 + (1001 * n);
+
+	for (size_t nb_coeff = 0; nb_coeff < (m / 2); nb_coeff++) {
+		r = (1 << 16);
+		while (r >= m) {
+			x = prbs23(x);
```

Comment:
`x` is undefined on the first run through

Reply:
Bit hard to spot, but it is initialized from `n` here: https://github.com/zephyrproject-rtos/zephyr/pull/73207/files/baee8fa2ba46c85186c38a3bda7ae07e873ab8be#diff-562bcad733f2caab662dfe65f9667d5e0ffe46fc9dea4a1aefdb857cc032fce7R349